dfastmi/batch/FileNameRetriever.py

```diff
@@ -29,119 +29,195 @@
     """
-    Extract the list of six file names from the configuration.
-    This routine is valid for version 1 configuration files.
-
-    Arguments
-    ---------
-    config : Optional[configparser.ConfigParser]
-        The variable containing the configuration (may be None for imode = 0).
-
-    Returns
-    -------
-    filenames : Dict[Any, Tuple[str,str]]
-        Dictionary of string tuples representing the D-Flow FM file names for
-        each reference/with measure pair. The keys of the dictionary vary. They
-        can be the discharge index, discharge value or a tuple of forcing
-        conditions, such as a Discharge and Tide forcing tuple.
+    Abstract file name retriever.
     """
-    filenames: Dict[Any, Tuple[str,str]]
-    key: Union[Tuple[float, int], float]
-    filenames = {}
-    for i in range(3):
-        qstr = "Q{}".format(i + 1)
-        if qstr in config:
-            reference = cfg_get(config, qstr, "Reference")
-            measure = cfg_get(config, qstr, "WithMeasure")
-            filenames[i] = (reference, measure)
-
-    return filenames
-
-
-def get_filenames_version2(
-    needs_tide: bool,
-    config: Optional[configparser.ConfigParser] = None,
-) -> Dict[Any, Tuple[str,str]]:
+    
+    @abstractmethod
+    def get_file_names(self, config : Optional[configparser.ConfigParser] = None) -> Dict[Any, Tuple[str,str]]:
+        """
+        Abstract method to get the filenames.
+
+        Arguments
+        ---------
+        config : Optional[configparser.ConfigParser]
+            The variable containing the configuration (may be None for imode = 0).
+
+        Returns
+        -------
+        filenames : Dict[Any, Tuple[str,str]]
+            Dictionary of string tuples representing the D-Flow FM file names for
+            each reference/with measure pair.
+        """
+    
+    def _cfg_get(self, config: configparser.ConfigParser, chap: str, key: str) -> str:
+        """
+        Get a single entry from the analysis configuration structure.
+        Raise clear exception message when it fails.
+
+        Arguments
+        ---------
+        config : Optional[configparser.ConfigParser]
+            The variable containing the configuration (may be None for imode = 0).
+        chap : str
+            The name of the chapter in which to search for the key.
+        key : str
+            The name of the key for which to return the value.
+
+        Raises
+        ------
+        Exception
+            If the key in the chapter doesn't exist.
+
+        Returns
+        -------
+        value : str
+            The value specified for the key in the chapter.
+        """
+        try:
+            return config[chap][key]
+        except:
+            raise Exception('Keyword "{}" is not specified in group "{}" of analysis configuration file.'.format(key,chap,))
+
+def get_filename_retriever(imode : int, config : configparser.ConfigParser, need_tides : bool) -> IFileNameRetriever:
```

Comment:
this is the factory right? Maybe create factory class and use a register function with dictionary and creator and a retrieve function to get correct creator or a default: 
(this is a out of my head example
```
class FileNameRetrieverFactory:
    """
    Class is used to register and get creation of AFileNameRetriever Objects
    """
    _creators = {}
    """Contains the AFileNameRetriever Objects creators to be used"""

    def __init__(self):
        self._creators = {}

    def register_creator(self, imode: int, creator: Callable[bool, AFileNameRetriever]):
        """Register creator function to create a AFileNameRetriever object."""
        self._creators[imode] = creator

    def generate(self, imode: int, needs_tide: bool) -> AFileNameRetriever:
        """Call the Constructor function to generate AFileNameRetriever object."""
        constructor = self._constructors.get(imode)
        if constructor:
            return constructor(needs_tide)
        else:
            return FileNameRetrieverUnsupported()

Reply:
and in the call class register mode when read / created (constructor / initializer core.py ?)
FileNameRetrieverFactory.register_creator(1, lambda needs_tide: FileNameRetrieverLegacy() )
FileNameRetrieverFactory.register_creator(2, lambda needs_tide: FileNameRetriever(needs_tide))